django/contrib/admin/views/main.py

```diff
@@ -384,3 +391,135 @@ def url_for_result(self, result):
+    """
+    View that handles requests from `.AdminForeignKeyWidget`.
+
+    The view only supports HTTP's GET method.
+    """
+    paginator_class = ForeignKeyJsonPaginator
+    paginate_by = 20
+    admin_site = None
+
+    def get(self, request, *args, **kwargs):
+        """
+        Return JSON search result based on a field identifier and a search term.
+
+        Returns:
+             (django.http.JsonResponse)::
+                {
+                    'results': [
+                        {
+                            'text': "foo",
+                            'id': 123
+                        }
+                    ],
+                    'more': true
+                }
+        """
+        field_id = kwargs.get('field_identifier', request.GET.get('field_identifier', None))
+        if not field_id:
+            raise PermissionDenied('No "field_identifier" provided.')
+        try:
+            self.app_label, self.model_name, self.field_name = field_id.split('.')
+        except ValueError:
+            raise PermissionDenied('Invalid "field_identifier".')
+        self.term = kwargs.get('field_identifier', request.GET.get('term', ''))
+        self.object_list = self.get_queryset()
+        context = self.get_context_data()
+        return JsonResponse({
+            'results': [
+                {
+                    'text': force_text(obj),
+                    'id': obj.pk,
+                }
+                for obj in context['object_list']
+            ],
+            'more': context['page_obj'].has_next()
+        })
+
+    def get_model(self):
+        return self.field.remote_field.model
+
+    model = property(get_model)
+
+    def get_field(self):
+        from django.apps import apps
+        try:
+            original_model = apps.get_model(self.app_label, self.model_name)
+            return original_model._meta.get_field(self.field_name)
+        except (LookupError, FieldDoesNotExist):
+            raise PermissionDenied('Invalid "field_identifier".')
+
+    field = property(get_field)
+
+    def get_queryset(self):
+        """Filter queryset based on the models' ``limit_choices_to`` and search term."""
+        queryset = super(ForeignKeyJsonView, self).get_queryset()
+        queryset = queryset.using(router.db_for_read(self.model))
+        model_field = self.get_field()
+        if hasattr(model_field, 'get_limit_choices_to'):
+            limit_choices_to = model_field.get_limit_choices_to()
+            if limit_choices_to is not None:
+                queryset = queryset.complex_filter(limit_choices_to)
+        result_set, search_use_distinct = self.model_admin.get_search_results(
+            self.request, queryset, self.term)
+        # Remove duplicates from results, if necessary
+        if search_use_distinct:
+            return result_set.distinct()
+        orderig = self.model_admin.get_ordering(self.request)
+        if orderig:
+            result_set = result_set.order_by(orderig)
```

Comment:
I think it also needs to be order_by(*ordering). (I get a TypeError) Would it be easier to use get_field_queryset() here?